File: cpp/src/exist.cpp
#include "types.hpp"

#include "util.hpp"

namespace filefs::detail {

std::uint8_t FileSystemState::stat(const std::string_view name) {
  if (name.empty() || !mounted()) {
    return 0U;
  }

  Block block{};
  const int len_n = static_cast<int>(name.size());
  std::uint32_t blockindex = 0U;
  int start = 0;
  if (name.front() == '/') {
    blockindex = 1U;
    start = 1;
  } else if (tmp.state == 0U) {
    blockindex = pwd_blockindex;
  } else {
    blockindex = tmp.pwd_blockindex;
  }

  std::array<char, BlockNameMaxSize + 2> s{};
  int slen = 0;
  for (int i = start; i < len_n; ++i) {
    if (name[static_cast<std::size_t>(i)] == '/') {
      if (slen == 0) {
        continue;
      }
      const std::uint32_t index = find_path_blockindex(
          blockindex, std::string_view{s.data(), static_cast<std::size_t>(slen)});
      if (index < 1U) {
        return 0U;
      }
      blockindex = index;
      slen = 0;
      continue;
    }
    s[static_cast<std::size_t>(slen++)] = name[static_cast<std::size_t>(i)];
    if (slen > static_cast<int>(BlockNameMaxSize)) {
      return 0U;
    }
  }
  if (slen == 0) {
    return 2U;
  }

  const std::string_view lastname{s.data(), static_cast<std::size_t>(slen)};
  if (!readblock(blockindex, block)) {
    return 0U;
  }

  const std::uint32_t stop_blockindex =
      b4_to_u32(std::span<const std::uint8_t, 4>(block.data() + 12 + 1 + 14 + 4, 4));
  const std::uint16_t offset =
      b2_to_u16(std::span<const std::uint8_t, 2>(block.data() + 12 + 1 + 14 + 4 + 4, 2));

  std::uint32_t index = blockindex;
  for (;;) {
    int k = BlockHead;
    for (int i = 0; i < BlockItemMaxCount; ++i) {
      const std::uint8_t state = block[static_cast<std::size_t>(k)];
      ++k;
      if (fixed_cstr(std::span<const std::uint8_t>(block.data() + k, BlockNameMaxSize)) != lastname) {
        k += static_cast<int>(BlockNameMaxSize) + 10;
        if (index == stop_blockindex && k + 1 >= static_cast<int>(offset)) {
          return 0U;
        }
        continue;
      }
      return (state & 0x01U) == 0U ? 2U : 1U;
    }

    index = b4_to_u32(std::span<const std::uint8_t, 4>(block.data() + 4, 4));
    if (index == 0U || !readblock(index, block)) {
      return 0U;
    }
  }
}

}  // namespace filefs::detail

namespace filefs {

bool FileSystem::file_exist(const std::string_view path) const {
  return state_ && state_->stat(path) == 1U;
}

bool FileSystem::dir_exist(const std::string_view path) const {
  return state_ && state_->stat(path) == 2U;
}

}  // namespace filefs

File: zig/src/exist.zig
const std = @import("std");
const types = @import("types.zig");
const util = @import("util.zig");
const block_mod = @import("block.zig");

pub fn stat(self: anytype, name: []const u8) u8 {
    if (name.len == 0 or self.fp == null) return 0;

    var block = util.blockZero();
    var blockindex: u32 = undefined;
    var start: usize = undefined;
    if (name[0] == '/') {
        blockindex = 1;
        start = 1;
    } else {
        blockindex = util.currentPwdBlockindex(self);
        start = 0;
    }

    var s: [types.BlockNameMaxSize + 2]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 2);
    var slen: usize = 0;
    var i: usize = start;
    while (i < name.len) : (i += 1) {
        if (name[i] == '/') {
            if (slen == 0) continue;
            const index = block_mod.findPathBlockindex(self, blockindex, s[0..slen]);
            if (index < 1) return 0;
            blockindex = index;
            slen = 0;
            continue;
        }
        s[slen] = name[i];
        slen += 1;
        if (slen > types.BlockNameMaxSize) return 0;
    }
    if (slen == 0) return 2;
    var lastname_buf: [types.BlockNameMaxSize + 1]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 1);
    @memcpy(lastname_buf[0..slen], s[0..slen]);
    const lastname = lastname_buf[0..slen];

    var b4: [4]u8 = undefined;
    var b2: [2]u8 = undefined;
    if (!block_mod.readblock(self, blockindex, block[0..])) return 0;
    @memcpy(b4[0..], block[12 + 1 + 14 + 4 .. 12 + 1 + 14 + 8]);
    const stop_blockindex = util.b4ToU32(b4[0..]);
    @memcpy(b2[0..], block[12 + 1 + 14 + 4 + 4 .. 12 + 1 + 14 + 4 + 6]);
    const offset = util.b2ToU16(b2[0..]);

    var index = blockindex;
    while (true) {
        var k: usize = types.BlockHead;
        var item: usize = 0;
        while (item < types.BlockItemMaxCount) : (item += 1) {
            const state = block[k];
            k += 1;
            @memset(s[0..], 0);
            @memcpy(s[0..types.BlockNameMaxSize], block[k .. k + types.BlockNameMaxSize]);
            k += types.BlockNameMaxSize;
            if (!std.mem.eql(u8, util.fixedCStr(s[0..types.BlockNameMaxSize + 1]), lastname)) {
                k += 10;
                if (index == stop_blockindex and k + 1 >= util.u16ToUsize(offset)) return 0;
                continue;
            }
            const dir_file = state & 0x01;
            if (dir_file == 0) return 2;
            return 1;
        }
        @memcpy(b4[0..], block[4..8]);
        index = util.b4ToU32(b4[0..]);
        if (index == 0) return 0;
        if (!block_mod.readblock(self, index, block[0..])) return 0;
    }
}

pub fn fileExist(self: anytype, filename: []const u8) bool {
    return stat(self, filename) == 1;
}

pub fn dirExist(self: anytype, pathname: []const u8) bool {
    return stat(self, pathname) == 2;
}
